### DISCUSSION TITLE
Reading measurements from flash on CLI

### DISCUSSION THREAD

**Discussion by oliver124810** [pvvx/ATC_MiThermometer#440] [Q&A]
Hi there,

first, thanks to all contributing developers for the great work!

My use case is to put the thermometers in various places, let them record measurements offline, and download the recorded data after a few days or weeks. For that, the custom firmware is great. I have flashed it successfully, and I can download the measurements from flash using GraphMemo.html as CSV.

Currently, downloading is quite cumbersome: start browser, open page, connect, download CSV, disconnect, repeat for every thermometer. Is there a CLI tool that can do this? If not, what would be the easiest way to create one? I have seen the Python interface, but it seems to be quite low-level

The ultimate goal would be to combine the gathered data in a database and display it graphically. I am considering Grafana and InfluxDB for that. If I can download the data with Python, it would be easy to upload it to InfluxDB, synchronously or asynchronously. Or is there a better option, knowing that the data is recorded offine?

Thank you for your help!

**Comment by JurySpb** [comment]
I'm also looking for a way to read saved data. General principle: You need to establish a connection and send a control command.
Unfortunately, there is not enough information about this or it is scattered. Here's what I found:
1) [ See](https://github.com/pvvx/ATC_MiThermometer#primary-service-uuid-0x1f10-characteristic-uuid-0x1f1f) - "Control function ID when connected"
2) View JavaScript code in "GraphMemo.html"
I'm using python on Linux. The first successes appeared.

**Comment by pvvx** [comment]
PrimaryService UUID16 **0x1F10**
Characteristic UUID16 **0x1F1F**

Enable Notifications.

Send/receive:
```
// Request history records
typedef struct {
	uint8_t		id;	// = 0x35 
	uint16_t	count; // number of records requested since last
} cmd_memo_t

// Reply to command
typedef struct {
	uint8_t		id;	// = 0x35
	uint16_t	count; // message number
	uint32_t	time; // Uinux timestamp (https://www.unixtimestamp.com/)
	int16_t		temp; // in 0.01 C
	uint16_t	humi; // in 0.01 %
	uint16_t	vbat; // in mV
} msg_memo_t

// End of transmission (size = 3 bytes)
typedef struct {
	uint8_t		id;	// = 0x35
	uint16_t	count; 
} msg_memo_end_t
```

When requested, it switches to “Connection interval” 20 ms, “Delay” 0.
~ 1133 package per second
When 2M PHY connected  it is twice as fast.

# Warning:
The ESP32 cannot receive this stream correctly.
When connected to an ESP, the thermometer's battery will be heavily discharged by transmitting duplicate packets.
Implementation of the program for ESP is possible only in C in ESP-IDF with Wi-Fi disabled, any debugging output in UART disabled, and static memory allocation, and disabling processing in any text form during reception. [Otherwise, ESP will not cope](https://github.com/pvvx/ATC_MiThermometer/discussions/328#discussioncomment-7838023).
For this reason, the communication interface is not published. Those who understand can always look at the sources, and not use Arduino to drain the batteries. Otherwise, the Lemings will copy this and scream about how quickly the batteries will drain.

**Comment by JurySpb** [comment]
Thank you!
```
// End of transmission (size = 3 bytes)
typedef struct {
	uint8_t		id;	// = 0x35
	uint16_t	count; 
} msg_memo_end_t
```
uint16_t	count - Always zero.
More information for Python. I'm using the bluepy library, not bluepy3. Bluepy3 doesn't work correctly for me.

**Reply by pvvx** [reply]
For https://github.com/pvvx/THB2 the interface is similar. But different UUIDs.
For https://github.com/pvvx/BLE_THSensor, the measurement memory size has been increased to 51943 pcs.

Likewise, this thread does not work on BT adapters using UART (RPi). There will also be excess consumption of device batteries. Only USB or PCI(E) adapters are suitable for BLE.


>  I'm using the bluepy library, not bluepy3. Bluepy3 doesn't work correctly for me.

This is another library for legacy Bluetooth 2014. Bluetooth 5.+ is not supported in them.
